### ISSUE TITLE
RFC: A(nother) uniform analyzer-driver interface

### ISSUE THREAD

**Issue by jaysachs** [kythe/kythe#3306]
# RFC: A(nother) uniform analyzer-driver interface

## Evolution

This proposal is an evolution of [RFC: A uniform analyzer-driver interface](#2801), and shamelessly incorporates large portions verbatim. The primary differences are:

1. Analyzers retrieve file contents from the filesystem and emit output to the filesystem; we essentially push the burden of virtualizing the filesystem from the analyzers (of which there are many, in many languages) to the drivers (which we expect there to be far fewer, in a small subset of languages).

2. We assume the the reduction of standardization on a single representation for compilation units as kzip to be accepted and implementation in-progress.

## Background

Kythe defines a loosely-coupled interface between build tools and analyzers. There are broadly two components of interest: _Extractors_ that consume source repositories and emit compilation records, and _analyzers_ that consume compilation records and emit analysis results (such as Kythe graph entries).

A **compilation record** in Kythe consists of a [kythe.proto.CompilationUnit][kcu], together with the contents of all the files referenced as required inputs to that unit. Work is underway to standardize on one format ([kzip][]), though two legacy formats ([indexpack][] and [kindex][]) still have some usage.

## Goals

There are two goals we'd like to address in this space:

1. To present analyzer authors with the most intuitive interface for obtaining artifacts for analysis (ultimately sourced from compilation records). Analyzers are often based on or derived from compilers, which typicially interact with file system APIs for obtaining sources.

2. To insulate analyzer authors from (changes in) the storage format of compilation records. Currently there is no way to implement an analyzer without supporting at least one of the existing file formats, and thus each analyzer must—unnecessarily, we argue—subsume some important implementation details of the analysis infrastructure.

We would like analyzers to have a simple and uniform interface to the rest of Kythe. Moreover, we would like that interface to depend as little as possible on concerns that are not relevant to the analyzer, such as how the compilation records are stored, how files are stored, cached, access-controlled, and fetched, and how the results of analysis are captured, stored, or propagated.

## Proposal

We propose that the interface between an analyzer and the rest of the Kythe infrastructure should be implemented as a text-based client-server protocol between the _analyzer_ (which is the client) and a _driver_ (which is the server). In summary, the interaction between analyzer and driver will work as follows:

1. The analyzer requests work from the driver, and the driver replies with a `AnalysisWorkUnit` message to be processed.

2. The analyzer requests any required files from the filesystem, based on the information provided in the `AnalysisWorkUnit` as well as from the contents of previously requested files. The driver is responsible for virtualizing the appropriate pathsin the filesystem, replying with the contents of those files, which based on the driver can be sourced from arbitrary locations.

3. The analyzer emits analysis outputs to a file specified in the `AnalysisWorkUnit`, and the driver disposes of them as appropriate.

The choice to use a text-based protocol is intentional. While all the Kythe data formats can be described by [protocol buffers][protobuf], we have found in practice that the quality of protobuf support varies widely among languages, and that in many cases the overhead of depending on the protobuf compiler and adding a code-generation step is a deterrent to contribution.

We bypass questions of efficiency by restricting the use of this text-based protocol to the analyzer-driver "control plane". For input source files, there is no encoding required; the virtualized files are represented as their original canonical byte streams. For output, we will permit the analyzer to indicate an encoding. Drivers are required to support at least JSON-encoding protobuf, and native protobuf wire-encoding. The efficiency gains and simplicity of native protobuf support in many analyzer host languages more than compensate for the small increase in complexity in supporting multiple output encodings.

### Protocol Details

More design details will be included in a separate document, but to a first approximation, we propose that:

* Analyzers and drivers should communicate via [JSON-RPC 2.0][jsonrpc2]. 

* Data that Kythe represents as protocol buffer definitions (such as the `CompilationUnit` and its affiliated details) will be encoded using the [canonical proto3 JSON encoding rules][p3json].

* Framing between client (analyzer) and server (driver) will follow the format used by the [Language Server Protocol (LSP) specification][lsp], which is length-prefixed and extensible.

## Discussion and Planning

If we obtain consensus on the basics of the proposal discussed above, we will need to do some work to update the existing analyzers to use this new protocol. Since, at the moment, most of the extant Kythe analyzers are either hosted in the Kythe repository or hosted internally at Google, it should be possible for the Kythe team to handle the conversion work directly.

Apart from converting the analyzers, we will also need to update some of the command-line tools that currently assume the indexer will be processing files directly. These changes should be self-contained to the Kythe repository, but we'll want to coordinate around releases to give our contributors time to respond and update.

## Potential flaw with this approach

One risk that needs to be evaluated is that when running an analyzer under virtualization, the execution of the analyzer binary itself may possibly request files, e.g. an analyzer written in Python loading packages. If there is overlap between the analyzer's filesystem dependencies and the files supplied by the driver from a compilation unit, the analyzer may fail to run, or run incorrectly. In particular this might seem to be a risk when the analyzer is requested to analyze its own source code.

[kcu]: https://github.com/google/kythe/blob/2e7b8085a4d84d2cddfef75b57bda08a61168aab/kythe/proto/analysis.proto#L85
[indexpack]: http://www.kythe.io/docs/kythe-index-pack.html
[kindex]: http://godoc.org/kythe.io/kythe/go/platform/kindex
[kzip]: http://www.kythe.io/docs/kythe-kzip.html
[protobuf]: https://developers.google.com/protocol-buffers
[jsonrpc2]: https://www.jsonrpc.org/specification
[p3json]: https://developers.google.com/protocol-buffers/docs/proto3#json
[lsp]: https://microsoft.github.io/language-server-protocol/specification

**Comment by creachadair** [comment]
> One risk that needs to be evaluated is that when running an analyzer under virtualization, the execution of the analyzer binary itself may possibly request files, e.g. an analyzer written in Python loading packages.

We already have this problem today, and I do not think either #2801 or #3306 makes it any worse. The boundary between a dynamic language runtime and its "static" dependencies is fluid and not strongly enforceable. Probably the best we can hope to do is to sandbox the analyzer and recommend—as we already have to the authors of internal analyzers—that retaining persistent state from one compilation record to the next is a bad choice.

The execution environment of an analyzer will always have multiple layers: Besides the operating system, the executable has all its own object code plus any shared libraries it needs. What this proposal adds on top is a swappable filesystem layer with some additional files in it. #2801 presents them via a JSON stream instead of the POSIX file API, but neither proposal can prevent the analyzer from carrying over data from one record to another.

So, in summary: If that's an issue, we already had it; and if it's not, we needn't worry about it here either.

I'll have some more comments on the rest after I've slept on it, but I wanted to respond to this while it was fresh in my mind.

**Comment by jaysachs** [comment]
I wasn't concerned, though maybe I should be, about leaking information from one analysis to the next. 

What I'm concerned about is something like this.

* analyzer binary depends on library file "/x/y/somelib.py"
* the compilation unit under analysis has "/x/y/somelib.py" as a required dependency file
* when running under file system virtualization, as the analyzer initializes, it requests "/x/y/somelib.py". Instead of obtaining it's build/deploy-time version of /x/y/somelib.py, it starts running the version in the compilation unit.

I understand this is largely hypothetical, as the analyzer would need to on-demand dynamically load needed libraries, and that a hermetic packaging of the analyzer binary would be sufficient to mitigate it. 

But I'm not sure I see how this potential issue already exists.

**Comment by creachadair** [comment]
> I wasn't concerned, though maybe I should be, about leaking information from one analysis to the next.

It depends what your threat model is. If the model requires us to isolate analyzers, then we have to run separate analyzers for each threat domain. That's obviously a separate issue, though (neither caused nor addressed by either proposal).
 
> What I'm concerned about is something like this.
> 
> * analyzer binary depends on library file "/x/y/somelib.py"
> * the compilation unit under analysis has "/x/y/somelib.py" as a required dependency file
> * when running under file system virtualization, as the analyzer initializes, it requests "/x/y/somelib.py". Instead of obtaining it's build/deploy-time version of /x/y/somelib.py, it starts running the version in the compilation unit.

Right. That _could_ in theory already happen today. Whether the front end of the compiler is wired up to pull from a driver via the service interface (which is how it works today) or whether it uses the stock "load from file" logic, it could easily have this problem. In practice they don't, though.

Still, you can't prevent it—as long as the same program has access to both its own source and the source under analysis in the same sandbox, this could occur. Such an analyzer would arguably be "poorly written" for not insulating the library invocation from the main program, but that's not really a new problem. 

For example: Nothing prevents the Java indexer from using its own class loader on the .class files we ship to be analyzed. It _doesn't_ do that, but it could. Sandboxing doesn't help: Even if you lock the analyzer out from its own source after initialization, you can't hide the compilation record from the analyzer and that's where the threat comes from.

I don't think it's likely to be a problem in practice.

**Comment by jaysachs** [comment]
> > I wasn't concerned, though maybe I should be, about leaking information from one analysis to the next.
> 
> It depends what your threat model is. If the model requires us to isolate analyzers, then we have to run separate analyzers for each threat domain. That's obviously a separate issue, though (neither caused nor addressed by either proposal).

We're on the same page. It is an issue, but unaffected by either proposal.

> Still, you can't prevent it—as long as the same program has access to both its own source and the source under analysis in the same sandbox, this could occur. Such an analyzer would arguably be "poorly written" for not insulating the library invocation from the main program, but that's not really a new problem.
I guess it just seems a little bit easier for an analyzer to be "poorly written". But ...

> I don't think it's likely to be a problem in practice.

I agree. Just wanted to be explicit that this was noticed and accepted.

**Comment by creachadair** [comment]
> I agree. Just wanted to be explicit that this was noticed and accepted.

Yes. I implicitly assumed that addressing this class of problems was out of scope when I wrote my original RFC, and I would argue it is here too. I agree it makes sense to be _ex_plicit about it, though.

**Comment by jaysachs** [comment]
> I'll have some more comments on the rest after I've slept on it, but I wanted to respond to this while it was fresh in my mind.

I would like to start on the more complete document, but am interested in further comments in case they materially affect things.

**Comment by creachadair** [comment]
> I would like to start on the more complete document, but am interested in further comments in case they materially affect things.

I actually don't have any major concerns. I imagine I'll have comments on the details, but I think the general approach is sound. 

With the filesystem approach, there will be some details to sort out, around how to lay out the container around the analyzer. The authors of analyzers (AoA) will need to know some rules about how paths are mapped from their original spelling—as recorded in the compilation record—to the actual filesystem. But that's all solvable with some work.

**Comment by creachadair** [comment]
The same tests (exactly) also fail in a container built on `ubuntu:latest`, so I am fair certain this is not due to some quirk of Alpine.

**Comment by jaysachs** [comment]
I have one more concern. While [gVisor](https://github.com/google/gvisor) appears to be a great fit for driver implementation, it is restricted to Linux only. I've done a bit of looking, and haven't found anything anywhere near as good for other OSes.  Sure, there are full virtualizations, but that seems heavyweight, and to support reading compilation unit contents from somewhere other than the actual filesystem, would require OS-level an FS. chroot might be another possibility but also doesn't provide any help with respect the filesystem, and absolutely requires root (I can't speak to the hypervisor virtualizations available).

Am I missing an obvious container-based virtualization solution for OSX? Should this be a concern?

**Comment by creachadair** [comment]
> I have one more concern. While [gVisor](https://github.com/google/gvisor) appears to be a great fit for driver implementation, it is restricted to Linux only. 

I think the only time that will matter is if there's an analyzer that will only run on macOS. Docker works fine on macOS but doesn't support hosting macOS in a container, unless you run a separate VM (e.g., via `docker-machine`). In general I would expect most analyzers to work on any OS, at which point it wouldn't matter whether they're running on the native local system.

In principle this could also affect Windows, but it's easy to run Windows containers in Docker from a Windows machine, unlike macOS containers on macOS.

So at least to a first approximation I don't think this problem is too bad. The ability to "containerize" analysis is part of the reason why the compilation record includes the file inputs _as given_. We talk about replay for _other_ analysis tools, but it also applies to the same analyzer on a different host. The latter is the dominant use case inside Google, for example—there the variation is among different Linux images rather than OS kernels. The net result is the same, though.

**Comment by jaysachs** [comment]
Thanks for pointing out that Docker on OS/X would in general work (aside from your caveat about OS/X-only indexers.) Is there a simple way when running Docker to intercept filesystem calls similar to how gVisor gofers work? It seems that we'd need plug into the virtual filesystem that is used to launch Docker; using containers naivelyI think we'd need to expand the entirety of the compilation unit into a dynamically-built docker image -- no "load on demand" for referenced file content.

Note that I'm not concerned about high-volume production use; having a Linux-primary solution seems satisfactory, at least for a while. But the use case I'm thinking about is smaller installations where someone checks out Kythe on their mac, and wants to build and serve an index. We'd be introducing friction into that process. 

Does this matter?

**Comment by creachadair** [comment]
> Thanks for pointing out that Docker on OS/X would in general work (aside from your caveat about OS/X-only indexers.) Is there a simple way when running Docker to intercept filesystem calls similar to how gVisor gofers work?

I'm not sure. I believe Docker for Mac is basically a wrapper around xhyve, so I guess technically the answer is "yes" but it might not be very easy. I was assuming, though, that once we've decided it's OK to ship the KZip to the analyzer in some form, it's mostly a matter of implementing a Docker storage driver that knows how to map the filesystem requests properly. At that point the KZip is basically a volume attached to the container. I haven't actually tried it, though, that's just how I've been thinking about it.

> It seems that we'd need plug into the virtual filesystem that is used to launch Docker; using containers naivelyI think we'd need to expand the entirety of the compilation unit into a dynamically-built docker image -- no "load on demand" for referenced file content.

Well, I think that mostly boils down to how much work you want to do on the storage driver. In theory you could completely virtualize the filesystem. (In practice that's kind of a pain, because getting filesystem semantics correct-enough to satisfy the underspecified corners of the POSIX file API can be surprisingly tricky).

> Note that I'm not concerned about high-volume production use; having a Linux-primary solution seems satisfactory, at least for a while. But the use case I'm thinking about is smaller installations where someone checks out Kythe on their mac, and wants to build and serve an index. We'd be introducing friction into that process.
> 
> Does this matter?

I think it matters to the extent that we shouldn't pick any design that explicitly _rules out_ a common platform. But if we can handle macOS, Windows, and Linux, I'd say the rest is probably negotiable.
